src/file-uploader/file-uploader.js

```diff
@@ -0,0 +1,167 @@
+
+    this.inputRef = React.createRef();
+    this.state = {isDragActive: false, acceptedFiles: [], rejectedFiles: []};
+  }
+
+  componentDidMount() {
+    window.addEventListener('dragover', this.onWindowDragOverOrDrop);
+    window.addEventListener('drop', this.onWindowDragOverOrDrop);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('dragover', this.onWindowDragOverOrDrop);
+    window.removeEventListener('drop', this.onWindowDragOverOrDrop);
+  }
+
+  updateFiles = (files: File[], event: SyntheticEvent<HTMLElement>) => {
+    if (this.props.accept) {
+      let nextFiles = {acceptedFiles: [], rejectedFiles: []};
+      nextFiles = files.reduce((acc, file) => {
+        if (checkFileType(file, this.props.accept)) {
+          acc.acceptedFiles.push(file);
+        } else {
+          acc.rejectedFiles.push(file);
+        }
+        return acc;
+      }, nextFiles);
+
+      if (this.props.onChange) {
+        this.props.onChange({...nextFiles, event});
+      }
+
+      this.setState({
+        acceptedFiles: [
+          ...this.state.acceptedFiles,
+          ...nextFiles.acceptedFiles,
+        ],
+        rejectedFiles: [
+          ...this.state.rejectedFiles,
+          ...nextFiles.rejectedFiles,
+        ],
+      });
+    } else {
+      const nextAcceptedFiles = {
+        acceptedFiles: [...this.state.acceptedFiles, ...files],
+      };
+
+      if (this.props.onChange) {
+        this.props.onChange({
+          ...nextAcceptedFiles,
+          rejectedFiles: this.state.rejectedFiles,
+          event,
+        });
+      }
+
+      this.setState(nextAcceptedFiles);
+    }
+  };
+
+  onInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
+    // spreading files to coerce to array because type of event.target.files is a 'FileList'
+    // https://github.com/facebook/flow/blob/v0.87.0/lib/dom.js#L55
+    this.updateFiles([...event.target.files], event);
+  };
+
+  onClick = (event: SyntheticEvent<HTMLButtonElement>) => {
+    if (!this.inputRef.current) {
+      return;
+    }
+
+    this.inputRef.current.value = '';
+    this.inputRef.current.click();
+  };
+
+  onDrop = (event: SyntheticDragEvent<HTMLElement>) => {
+    event.preventDefault();
+    this.setState({dragging: false});
+    const files = event.dataTransfer ? event.dataTransfer.files : undefined;
+
+    if (files) {
+      this.updateFiles(files, event);
+    }
+  };
+
+  onDragEnter = (event: SyntheticDragEvent<HTMLElement>) => {
+    this.setState({isDragActive: true});
+  };
+
+  onDragLeave = (event: SyntheticDragEvent<HTMLElement>) => {
+    this.setState({isDragActive: false});
+  };
+
+  onDragOver = (event: SyntheticDragEvent<HTMLElement>) => {
+    event.preventDefault();
+  };
+
+  onWindowDragOverOrDrop = (event: SyntheticDragEvent<HTMLElement>) => {
+    event.preventDefault();
+  };
+
+  render() {
+    return (
+      <Root>
+        <ContentContainer>
+          <FileDragAndDrop
+            onDrop={this.onDrop}
```

Comment:
Would it make sense to make the whole area droppable? So the onDrag handlers will be on a container component.

Reply:
i've changed the markup here, so not sure if the comment still applies. but i wanted to make an outer wrapper that contains all of the elements, and just the area with the dashed border accepts drops. i'd like to display this in the design meeting, so will accept feedback on how it should operate